modeling/algorithm/local-weighted-slice-algorithm.go

```diff
@@ -0,0 +1,202 @@
+		localGroup.Label = zoneName
+		// this local sliceGroup should only receive traffic from current zone,
+		// this map tracks weights of traffic from different zones to this
+		// sliceGroup
+		localGroup.ZoneTrafficWeights = map[string]float64{zoneName: 1.0}
+		// this map stores composition of this sliceGroup, zoneName - number of
+		// endpoints from zoneName
+		localGroup.Composition = map[string]types.WeightedEndpoints{}
+
+		// calculate expected number of endpoints based on the proportion of
+		// nodes in this zone
+		expectedEndpoints := zone.NodesRatio * float64(region.TotalEndpoints)
+		// deviation: a negative value means need more endpoints from other
+		// sliceGroups, a positive value means need give out endpoints to other
+		// sliceGroups
+		deviation := float64(zone.Endpoints) - expectedEndpoints
+		// intDeviation is dealt same way as original local-slice algorithm that
+		// directly 'borrows' or 'sends' endpoints among zones
+		intDeviation := int(deviation)
+		if intDeviation == 0 {
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else if intDeviation > 0 {
+			endpointsAvailable.push(endpointDeviation{name: zoneName, deviation: intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: zone.Endpoints, Weight: 1}
+		}
+		sliceGroups[zoneName] = localGroup
+
+		// push 0.xx deviation into corresponding lists
+		floatDeviation := deviation - float64(intDeviation)
+		if floatDeviation > 0 {
+			// Intentionally set deviation to default 0, this means the zone
+			// itself needs float part from this endpoint otherwise means the
+			// endpoints will be shared by other zones only
+			weightedEndpointsAvailable.push(endpointDeviation{name: zoneName, weight: floatDeviation})
+		} else if floatDeviation < 0 {
+			weightedEndpointsNeeded.push(endpointDeviation{name: zoneName, deviation: 1, weight: -floatDeviation})
+		}
+	}
+
+	err := alg.balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, &weightedEndpointsAvailable, &weightedEndpointsNeeded, sliceGroups)
+	return sliceGroups, err
+}
+
+// balanceSliceGroups distributes endpoints from zones with extra endpoints to
+// EndpointSliceGroups for zones with insufficient endpoints.
+func (alg LocalWeightedSliceAlgorithm) balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, weightedEndpointsAvailable *endpointsList, weightedEndpointsNeeded *endpointsList, sliceGroups map[string]types.EndpointSliceGroup) error {
+
+	for _, receiveZone := range endpointsNeeded.byZone {
+		// No more 'int' endpoints available but are still 'int' endpoints in
+		// need, we push the needed entpoints into weighted list and deal them
+		// as float endpoints
```

Comment:
Not quite sure I'm understanding this correctly, but maybe something like:
```
There are no more full endpoints available, but this zone still needs endpoints. Push the needed endpoints into a weighted list and deal with them as partial endpoints.
```

Reply:
done